File: languages/zig/cross-language-features.zig
const std = @import("std");
const xl = @import("willyhorizont/runtime/xl.zig");

pub fn main(init: std.process.Init) !void {
    xl.init_runtime(init.gpa, init.io);

    // # ' -- 1. support lambda as value, or has workaround
    const say_hello = xl.init_lambda(.{}, struct {
        fn call(ctx: anytype, va: []const xl.Type) xl.Type {
            _ = ctx;
            var itr = xl.iter(va);
            const callback = itr.next();
            xl.print(.{xl.init_string("hello")});
            _ = callback.call(.{});
            return xl.none;
        }
    }.call);
    defer say_hello.deinit();
    _ = say_hello.call(.{xl.init_lambda(.{}, struct {
        fn call(ctx: anytype, va: []const xl.Type) xl.Type {
            _ = ctx;
            _ = va;
            xl.print(.{xl.init_string("world")});
            return xl.none;
        }
    }.call)});
    const create_multiplier = xl.init_lambda(.{}, struct {
        fn call(ctx_aa: anytype, va_aa: []const xl.Type) xl.Type {
            _ = ctx_aa;
            var itr_aa = xl.iter(va_aa);
            const aa = itr_aa.next();
            return xl.init_lambda(.{ .aa = aa }, struct {
                fn call(ctx_bb: anytype, va_bb: []const xl.Type) xl.Type {
                    var itr_bb = xl.iter(va_bb);
                    const bb = itr_bb.next();
                    return xl.init_int(ctx_bb.aa.to_int() * bb.to_int());
                }
            }.call);
        }
    }.call);
    defer create_multiplier.deinit();
    const multiply_by_two = create_multiplier.call(.{xl.init_int(2)});
    defer multiply_by_two.deinit();
    const multiply_by_eight = create_multiplier.call(.{xl.init_int(8)});
    defer multiply_by_eight.deinit();
    xl.print(.{ "multiply_by_two(8): ", xl.json_stringify(multiply_by_two.call(.{xl.init_int(8)}), .{}) });
    xl.print(.{ "multiply_by_eight(4): ", xl.json_stringify(multiply_by_eight.call(.{xl.init_int(4)}), .{}) });
    xl.print(.{ "multiply_by_two(8): ", xl.json_stringify(multiply_by_two.call(.{xl.init_int(8)}), .{}) });

    // # ' -- 2. support dynamic-typed value, or has workaround
    const xl_list = xl.init_list(.{
        xl.none,
        xl.init_bool(true),
        xl.init_bool(false),
        xl.init_string("foo"),
        xl.init_int(0),
        xl.init_int(-123),
        xl.init_float(123.789),
        xl.init_float(-123.789),
        xl.init_list(.{ xl.init_int(1), xl.init_int(2), xl.init_int(3) }),
        xl.init_dict(.{.{ "foo", xl.init_string("bar") }}),
        xl.init_lambda(.{}, struct {
            fn call(ctx: anytype, va: []const xl.Type) xl.Type {
                _ = ctx;
                var itr = xl.iter(va);
                const aa = itr.next();
                const bb = itr.next();
                return xl.init_int(aa.to_int() * bb.to_int());
            }
        }.call),
    });
    defer xl_list.deinit();
    xl.print(.{ "xl_list: ", xl.json_stringify(xl_list, .{}) });
    xl.print(.{ "xl_list: ", xl.json_stringify(xl_list, .{ .pretty = true }) });
    const xl_dict = xl.init_dict(.{
        .{ "xl_none", xl.none },
        .{ "xl_bool_true", xl.init_bool(true) },
        .{ "xl_bool_false", xl.init_bool(false) },
        .{ "xl_string", xl.init_string("foo") },
        .{ "xl_int_positive", xl.init_int(0) },
        .{ "xl_int_negative", xl.init_int(-123) },
        .{ "xl_float_positive", xl.init_float(123.789) },
        .{ "xl_float_negative", xl.init_float(-123.789) },
        .{ "xl_list", xl.init_list(.{ xl.init_int(1), xl.init_int(2), xl.init_int(3) }) },
        .{ "xl_dict", xl.init_dict(.{.{ "foo", xl.init_string("bar") }}) },
        .{ "xl_lambda", xl.init_lambda(.{}, struct {
            fn call(ctx: anytype, va: []const xl.Type) xl.Type {
                _ = ctx;
                var itr = xl.iter(va);
                const aa = itr.next();
                const bb = itr.next();
                return xl.init_int(aa.to_int() * bb.to_int());
            }
        }.call) },
    });
    defer xl_dict.deinit();
    xl.print(.{ "xl_dict: ", xl.json_stringify(xl_dict, .{}) });
    xl.print(.{ "xl_dict: ", xl.json_stringify(xl_dict, .{ .pretty = true }) });
}

File: languages/java/cross-language-features.java
import willyhorizont.runtime.Xl;

public class Main {
    public static void main(String[] args) {
        /*
        1. support lambda as value, or has workaround
        */
        Xl sayHello = Xl.initLambda((Xl... va) -> {
            Xl itr = Xl.iter(va);
            Xl callback = itr.next();
            System.out.println("hello");
            callback.call();
            return Xl.NONE;
        });
        sayHello.call(Xl.initLambda((Xl... va) -> {
            System.out.println("world");
            return Xl.NONE;
        }));
        Xl createMultiplier = Xl.initLambda((Xl... vaAa) -> {
            Xl itrAa = Xl.iter(vaAa);
            Xl aa = itrAa.next();
            return Xl.initLambda((Xl... vaBb) -> {
                Xl itrBb = Xl.iter(vaBb);
                Xl bb = itrBb.next();
                return Xl.from(aa.toInt() * bb.toInt());
            });
        });
        Xl multiplyByTwo = createMultiplier.call(Xl.from(2));
        System.out.println("multiply_by_two(10): " + multiplyByTwo.call(Xl.from(10)));
        Xl multiplyByEight = createMultiplier.call(Xl.from(8));
        System.out.println("multiply_by_eight(4): " + multiplyByEight.call(Xl.from(4)));
        System.out.println("multiply_by_two(8): " + multiplyByTwo.call(Xl.from(8)));

        /*
        2. support dynamic-typed value, or has workaround
        */
        Xl xlList = Xl.initList(
            Xl.NONE,
            true,
            false,
            "foo",
            0,
            -123,
            123.789,
            -123.789,
            Xl.initList(1, 2, 3),
            Xl.initDict(Xl.initPair("foo", "bar")),
            Xl.initLambda((Xl... va) -> {
                Xl itr = Xl.iter(va);
                Xl aa = itr.next();
                Xl bb = itr.next();
                return Xl.from(aa.toInt() * bb.toInt());
            })
        );
        System.out.println(Xl.jsonStringify(xlList));
        System.out.println(Xl.jsonStringify(xlList, Xl.initPair("pretty", true)));
        Xl xlDict = Xl.initDict(
            Xl.initPair("xl_none", Xl.NONE),
            Xl.initPair("xl_bool_true", true),
            Xl.initPair("xl_bool_false", false),
            Xl.initPair("xl_string", "foo"),
            Xl.initPair("xl_int_positive", 0),
            Xl.initPair("xl_int_negative", -123),
            Xl.initPair("xl_float_positive", 123.789),
            Xl.initPair("xl_float_negative", -123.789),
            Xl.initPair("xl_list", Xl.initList(1, 2, 3)),
            Xl.initPair("xl_dict", Xl.initDict(Xl.initPair("foo", "bar"))),
            Xl.initPair("xl_lambda", Xl.initLambda((Xl... va) -> {
                Xl itr = Xl.iter(va);
                Xl aa = itr.next();
                Xl bb = itr.next();
                return Xl.from(aa.toInt() * bb.toInt());
            }))
        );
        System.out.println(Xl.jsonStringify(xlDict));
        System.out.println(Xl.jsonStringify(xlDict, Xl.initPair("pretty", true)));
    }
}
